pkg/webhook/clustersetbinding/validating_webhook.go

```diff
@@ -0,0 +1,145 @@
+func (a *ManagedClusterSetBindingValidatingAdmissionHook) Validate(admissionSpec *admissionv1beta1.AdmissionRequest) *admissionv1beta1.AdmissionResponse {
+	klog.V(4).Infof("validate %q operation for object %q", admissionSpec.Operation, admissionSpec.Object)
+
+	status := &admissionv1beta1.AdmissionResponse{}
+
+	// only validate the request for ManagedClusterSetBinding
+	if admissionSpec.Resource.Group != "cluster.open-cluster-management.io" ||
+		admissionSpec.Resource.Resource != "managedclustersetbindings" {
+		status.Allowed = true
+		return status
+	}
+
+	switch admissionSpec.Operation {
+	case admissionv1beta1.Create, admissionv1beta1.Update:
+		return a.validateRequest(admissionSpec)
+	default:
+		status.Allowed = true
+		return status
+	}
+}
+
+// Initialize is called by generic-admission-server on startup to setup initialization that ManagedClusterSetBinding webhook needs.
+func (a *ManagedClusterSetBindingValidatingAdmissionHook) Initialize(kubeClientConfig *rest.Config, stopCh <-chan struct{}) error {
+	var err error
+	a.kubeClient, err = kubernetes.NewForConfig(kubeClientConfig)
+	if err != nil {
+		return err
+	}
+
+	a.clusterClient, err = clusterclient.NewForConfig(kubeClientConfig)
+	return err
+}
+
+// validateCreateRequest validates the create/update operation of ManagedClusterSetBinding
+func (a *ManagedClusterSetBindingValidatingAdmissionHook) validateRequest(request *admissionv1beta1.AdmissionRequest) *admissionv1beta1.AdmissionResponse {
+	status := &admissionv1beta1.AdmissionResponse{}
+
+	binding := &clusterv1alpha1.ManagedClusterSetBinding{}
+	if err := json.Unmarshal(request.Object.Raw, binding); err != nil {
+		status.Allowed = false
+		status.Result = &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+			Message: err.Error(),
+		}
+		return status
+	}
+
+	// check if target cluster set is not specified
+	if len(binding.Spec.ClusterSet) == 0 {
+		status.Allowed = false
+		status.Result = &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+			Message: "Spec.ClusterSet is required",
+		}
+		return status
+	}
+
+	// check if the request user has permission to bind the target cluster set
+	return a.allowBindingToClusterSet(binding.Spec.ClusterSet, request.UserInfo)
+}
+
+// allowBindingToClusterSet checks if the user is able to bind a cluster set
+func (a *ManagedClusterSetBindingValidatingAdmissionHook) allowBindingToClusterSet(clusterSetName string, userInfo authenticationv1.UserInfo) *admissionv1beta1.AdmissionResponse {
+	status := &admissionv1beta1.AdmissionResponse{}
+
+	extra := make(map[string]authorizationv1.ExtraValue)
+	for k, v := range userInfo.Extra {
+		extra[k] = authorizationv1.ExtraValue(v)
+	}
+
+	sar := &authorizationv1.SubjectAccessReview{
+		Spec: authorizationv1.SubjectAccessReviewSpec{
```

Comment:
this should be documented in your api godoc.  apologies if you did this after I reviewed before.

Reply:
Doc is added